### PR TITLE
Bbg support ios simulator

### PR DESCRIPTION
**Pull request by BruceBGordon**
Each worker discovers the local ios simulator devices at startup and adds them to device-manager's devices Jobs can request a simulator. Docs updated.

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -25,6 +26,12 @@ if (argv.config && argv.config === true) {
     throw new Error('--config defined without any following configFile')
 }
 
+if (argv.saveTempFiles) {
+    logger.i('--saveTempFiles option: Guerilla temp files in the OS temp directory will not be deleted at process exit.');
```
Comment (dbgrandi) [inline lib/config.js:30]: `saveTempFiles` is a slightly ambiguous name without the description. They are always saved. :smile:

File: lib/context.js [hunk 2]
```diff
@@ -17,7 +17,10 @@ function Context (executor, job, result, device) {
 	self.bin_dir = path.join(__rootdir, 'bin');
 	self.output_dir = result.output_dir;
 	self.working_dir = job.getWorkingDir();
-	if (device) self.device_identifier = device.identifier;
+	if (device) {
+		self.device_identifier = device.identifier;
+		self.device = device;
```
Comment (dbgrandi) [inline lib/context.js:21]: Why hold `self.device_identifier` when you could just use `self.device.identifier`?

File: lib/device-manager.js [hunk 3]
```diff
@@ -6,29 +7,174 @@
...
+ * @param line String of the form 'iPhone 4s (05686243-BA24-44CE-94D9-D6823BF96E46) (Shutdown) {(unavailable, ...)}'
+ * @returns {{tag: string}}
+ */
+
+/**
+ *
+ * @param oslevel String
+ * @param line String to parse
+ * @returns Device-like JSON object
+ */
```
Comment (dbgrandi) [inline lib/device-manager.js:33]: Looks like you have the doc comments repeated here. Should merge.

File: lib/device-manager.js [hunk 4]
```diff
@@ -6,29 +7,174 @@
...
+function createDevice(oslevel, line) {
+	if (line.indexOf('unavailable') !== -1) throw new Error('createDevice-fed an unavailable device:' + oslevel + ':' +  line);
+	var openparenIndex = line.indexOf('(');
```
Comment (dbgrandi) [inline lib/device-manager.js]: capitalization on `openparenIndex`

File: lib/device-manager.js [hunk 5]
```diff
@@ -6,29 +7,174 @@
...
+		platform: 'ios',
+        identifier: identifier,
+		name: 'ios-simulator ' + name + ' (' + osVersion + ')',
+		OS: oslevel,
+		simulator: true,
+		destination: 'OS=' + osVersion + ',name=' + name
+	};
+}
+
+function isHeader(line) {
```
Comment (dbgrandi) [inline lib/device-manager.js:55]: Does this method need a comment?

File: lib/device-manager.js [hunk 6]
```diff
@@ -6,29 +7,174 @@
...
+	var snippet =  line.slice(0,2);
+	return (snippet === '--' || snippet === '==');
+}
+
+/**
+ * A line oriented Stream over Simctl output from running $xcrun simctl list
+ * @param array {Array.<String>}
+ * @returns {{next: Function, advanceTillMatch: Function, hasMore: Function, nextValue: Function, nextValueAsOSLevel: Function, peekHasNextLineEntry: Function, peekIsHeader: Function, peekIsDoubleEquals: Function, peek: Function}}
+ */
+function makeSimctlStream(array){
```
Comment (dbgrandi) [inline lib/device-manager.js]: missing space before '{'

Comment (dbgrandi) [inline lib/device-manager.js]: I'd also avoid using `array` as an arg name. Very close to `Array`, which I'm assuming should be treated like a reserved word.

File: lib/config.js (hunk 1 above)
Comment (BruceBGordon) [inline lib/config.js:30]: I'm open to a better name.

File: lib/device-manager.js [hunk 7]
```diff
@@ -6,29 +7,174 @@
...
-	for (var i = 0; i < this.devices.length; i++) {
-		if (this.devices[i].tag == tag)
-			return this.devices[i];
-	}
 
-	return null;
+
+DeviceManager.prototype.findByTag = function (tag) {
+	var result = this.devicesByTag.get(tag);
+	return result ? result : null;
```
Comment (dbgrandi) [inline lib/device-manager.js:182]: Is there a JS shorthand for this, like `return result ?: null`

File: lib/context.js (hunk 2 above)
Comment (BruceBGordon) [inline lib/context.js:21]: ok, the cure is on the way. Hope it's better than the disease.

File: lib/device-manager.js (hunk 7 above)
Comment (BruceBGordon) [inline lib/device-manager.js:182]: Not that I know.

**Comment by BruceBGordon** [comment]
addressed code review points with new  commit

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by dbgrandi** [comment]
Wow, that `device.identifier` is everywhere.

Should absolutely get Travis setup and set your avatar. :+1:

**Pull request merged**